olmo/train.py

```diff
@@ -753,10 +823,31 @@ def train_batch(self, batch: Dict[str, Any]) -> Tuple[torch.Tensor, Optional[tor
                         assert z_batch_loss is not None
                         z_batch_loss += z_loss.detach()
 
+                if self.model.config.block_type == BlockType.moe:
+                    if self.model.config.moe_zloss_weight is not None:
+                        lb_loss, moe_z_loss = batched_load_balancing_loss(self.moe_args)
+                        lb_loss = lb_loss / len(micro_batches)
+                        moe_z_loss = moe_z_loss / len(micro_batches)
+                    else:
+                        lb_loss = batched_load_balancing_loss(self.moe_args) / len(micro_batches)
+                    if self.model.config.moe_log_expert_assignment:
+                        if self.model.config.moe_zloss_weight:
+                            tokens_per_expert, _, _ = zip(*get_load_balancing_loss())
+                        else:
+                            tokens_per_expert, _ = zip(*get_load_balancing_loss())
+                        expert_assignments += torch.stack(tokens_per_expert, dim=0).cpu()
+                    clear_load_balancing_loss()
+                    if self.model.config.moe_loss_weight:
+                        loss += lb_loss
+                    if self.model.config.moe_zloss_weight:
+                        loss += moe_z_loss
```

Comment:
Where do those weights get applied?

Reply:
It looks like they are just used as a boolean flag.

Reply:
In megablocks via the MoE config (`"moe_loss_weight": self.model.config.moe_loss_weight,`)